.github/workflows/qa.yml

```diff
@@ -1,26 +1,113 @@
+                        echo
+                        exit 1
+                    fi
+
+            -
+                name: No .yml, use .yaml
+                if: "always() && steps.checkout.outcome == 'success'"
+                run: |
+                    YMLS=$(find * -name '*.yml')
+                    if [[ "" != "$YMLS" ]]; then
+                        echo "::error::*.yml files should renamed to *.yaml"
+                        echo "Found $YMLS"
+                        echo
+                        exit 1
+                    fi
+
+            -
+                name: No .gitkeep, use .gitignore
+                if: "always() && steps.checkout.outcome == 'success'"
+                run: |
+                    GITKEEPS=$(find * -name .gitkeep)
+                    if [[ "" != "$GITKEEPS" ]]; then
+                        echo "::error::.gitkeep files should be renamed to .gitignore"
+                        echo "Found $GITKEEPS"
+                        echo
+                        exit 1
+                    fi
+
+            -
+                name: 4 spaces indentation
+                if: "always() && steps.checkout.outcome == 'success'"
+                run: |
+                    ERRORS=$(find * -name '*.yaml' -or -name '*.json' \
+                        | xargs -n1 grep -P -H -n -v '^((    )*[^ \t]|$)' \
+                        | cut -d: -f1-2 \
+                        | sed 's/\(.*\):\([0-9]*\)$/\\n::error file=\1,line=\2::Indendation must be a multiple of 4 spaces/' || true)
+
+                    if [[ "" != "$ERRORS" ]]; then
+                        echo -e $ERRORS
+                        echo
+                        exit 1
+                    fi
+
+            -
+                name: manifest.json is found
+                if: "always() && steps.checkout.outcome == 'success'"
+                run: |
+                    MISSING=$(find * -mindepth 2 -maxdepth 2 -type d '!' -exec test -f '{}/manifest.json' ';' -print)
+
+                    if [[ "" != "$MISSING" ]]; then
+                        echo "::error::Recipes must define a \"manifest.json\" file"
+                        echo "File not found in $MISSING"
+                        echo
+                        exit 1
+                    fi
+
+            -
+                name: JSON files are valid
+                if: "always() && steps.checkout.outcome == 'success'"
+                run: |
+                    ERRORS=$(find * -name '*.json' | parallel -j+3 -i{} -n1 bash -c 'ERR=$(python -mjson.tool {} 2>&1 1> /dev/null) || echo \\n::error file={},line=`echo "${ERR#*: line }" | cut -d" " -f 1`::${ERR%%: line *}')
+
+                    if [[ "" != "$ERRORS" ]]; then
+                        echo -e $ERRORS
+                        echo
+                        exit 1
+                    fi
+
+            -
+                name: YAML files are valid
+                if: "always() && steps.checkout.outcome == 'success'"
```

Comment:
this should check the state of the install of the tools, not of the checkout, as it relies on this setup step too

Reply:
Not sure we care: the install step is not expected to fail, and if it does, this one will too.